### PR TITLE
miri: accept extern types in structs if they are the only field

### PR DESCRIPTION
**Pull request by RalfJung**
Fixes https://github.com/rust-lang/rust/issues/55541

Cc @oli-obk @eddyb https://github.com/rust-lang/rust/issues/43467

### CODE REVIEW

**Comment by eddyb** [comment]
I would prefer it if this was based on offset being `0`, *not* field count.

**Comment by RalfJung** [comment]
Like this?

**Comment by eddyb** [comment]
I'm not sure the alignment condition is needed, since `0` is aligned to infinity, but I don't think it hurts.

File: src/test/ui/consts/const-eval/issue-55541.rs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+#![feature(extern_types, const_transmute)]
+
+extern "C" {
+  pub type ExternType;
+}
+unsafe impl Sync for ExternType {}
+
+#[repr(transparent)]
+pub struct Wrapper(ExternType);
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply src/test/ui/consts/const-eval/issue-55541.rs]: Done.

File: src/librustc_mir/interpret/place.rs [hunk 2]
```diff
@@ -354,7 +354,8 @@ where
         let (meta, offset) = if field_layout.is_unsized() {
             // re-use parent metadata to determine dynamic field layout
             let (_, align) = self.size_and_align_of(base.meta, field_layout)?
-                .expect("Fields cannot be extern types");
+                // If this is an extern type, we fall back to its static size and alignment.
+                .unwrap_or_else(|| base.layout.size_and_align());
```
Comment (eddyb) [inline src/librustc_mir/interpret/place.rs]: Interestingly, we can make `size_and_align_of` error, but have a separate `align_of` that works even for `extern type`. Only the alignment is used here, anyway!

Comment (eddyb) [reply src/librustc_mir/interpret/place.rs]: Oh, you can also just move the `Option` to be around the size, but not the alignment.

Comment (RalfJung) [reply src/librustc_mir/interpret/place.rs]: I could, but I'd rather wait for the decision to treat the `size_of_val` and `align_of_val` intrinsics asymmetrically before doing that.

`extern type` will likely get alignment 1 currently, which is likely not actually correct for the type they are opaquely representing, so I am a bit worried about unilaterally implementing such behavior.

Comment (eddyb) [reply src/librustc_mir/interpret/place.rs]: You can't access that alignment *without* first having a reference, so having it too high would be a problem (the assumption might not be guaranteed by what you're FFI-ing to), but having it too low wouldn't break anything (other than field offsets).

So the way I see it, the alignment is only a lower bound, and `1` is fine, unless you want to use it as an aligned field in a struct.

Anyway, my concern is also kind of a confusing code thing.
You could also add a `.map(|(_, align)| align)` before `unwrap_or_else`, which would mean the code and the comment wouldn't mention the "size" of the `extern type`.

Comment (RalfJung) [reply src/librustc_mir/interpret/place.rs]: > You can't access that alignment without first having a reference, so having it too high would be a problem (the assumption might not be guaranteed by what you're FFI-ing to), but having it too low wouldn't break anything (other than field offsets).

*Oh* that's a good point. We cannot create such references ourselves anyway as we cannot have data of that type.

Still I think miri shouldn't move ahead of the rest of the decision process here.

I will implement your `map` suggestion.

Comment (RalfJung) [reply src/librustc_mir/interpret/place.rs]: I implemented what you suggested.

**Review by eddyb — APPROVED** [review]

File: src/librustc_mir/interpret/place.rs [hunk 3]
```diff
@@ -353,8 +353,13 @@ where
...
             // re-use parent metadata to determine dynamic field layout
-            let (_, align) = self.size_and_align_of(base.meta, field_layout)?
-                .expect("Fields cannot be extern types");
+            let align = self.size_and_align_of(base.meta, field_layout)?
+                .map(|(_, align)| align)
+                // If this is an extern type, we fall back to its static alignment.
+                // FIXME: Once we have made decisions for how to handle size and alignment
+                // of `extern type`, this should be adapted.  It is just a temporary hack
+                // to get some code to work that probably ought to work.
+                .unwrap_or_else(|| base.layout.align);
```
Comment (RalfJung) [inline src/librustc_mir/interpret/place.rs]: Should this really be `base.layout`, not `field_layout`?

**Comment by RalfJung** [comment]
Hm actually I am not entirely sure I got this right, will this error out appropriately when there are other fields? It will error out in `size_and_align_of`, but what about `mplace_field`? It does not do any such check.

**Comment by eddyb** [comment]
@RalfJung Ah, yeah, you probably want it in both. This is why decoupling alignment from size should help - size and offset computation are similar/equivalent, whereas alignment is simpler.

**Comment by RalfJung** [comment]
Offset computation of a field only uses the field's alignment, though (and the `offsets` data in the layout).

**Comment by RalfJung** [comment]
I fixed that, I think.

**Comment by nikic** [comment]
It looks like #55624 is trying to fix the same issue, but in a different way (specific to repr(transparent)). Could it be that this PR subsumes #55624?

**Comment by RalfJung** [comment]
WTF?
```
test time::tests::instant_duration_panic ... ok
test time::tests::instant_elapsed ... ok
thread '<unnamed>' panicked at 'assertion failed: a - Duration::new(0, 100) <= b', libstd\time.rs:507:9
test time::tests::instant_math ... FAILED
test time::tests::instant_monotonic ... ok
test time::tests::since_epoch ... ok
test time::tests::system_time_elapsed ... ok
test time::tests::system_time_math ... ok
test sync::mpsc::tests::stress_recv_timeout_shared ... ok
test thread::tests::test_park_timeout_unpark_called_other_thread ... ok
test sync::mpsc::tests::stress_recv_timeout_two_threads ... ok

failures:

failures:
    time::tests::instant_math
```
Most jobs came out green, and this one has a failure that makes no sense. What is going on?

Unfortunately the error log has so many "something panicked" messages, it is impossible to tell if any one of them has anything to do with this test failure.

**Comment by RalfJung** [comment]
Oh, this test actually measures the time passing between two calls to `Instant::now()`:
```rust
    fn instant_math() {
        let a = Instant::now();
        let b = Instant::now();
        let dur = b.duration_since(a);
        assert_almost_eq!(b - dur, a);
        assert_almost_eq!(a + dur, b);

        let second = Duration::new(1, 0);
        assert_almost_eq!(a - second + second, a);
    }
```

I am going to assume this is spurious and the test runner was just really slow. Cc @rust-lang/infra

@bors retry

**Comment by RalfJung** [comment]
I am doing r- because I'd like my other PR https://github.com/rust-lang/rust/pull/55716 to land first. Hope that's okay.

@bors r-

**Comment by kennytm** [comment]
@RalfJung  You could just make 55716 p=1 or something

**Comment by RalfJung** [comment]
That would make it overtake other people's PRs as well, which did not seem appropriate.

**Comment by kennytm** [comment]
You could add a `p` to those people's PR as well 😂. In the current queue the only PR above 55716 is #55356.

**Comment by RalfJung** [comment]
The other PR landed.

@bors r=eddyb

**Comment by RalfJung** [comment]
It's that timing test again:
```
failures:
    time::tests::instant_math
```

@bors retry

@rust-lang/infra maybe the permitted time difference should be increased to 500ms or so?

**Comment by kennytm** [comment]
@bors retry 

```
error: failed to remove file `C:\projects\rust\build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\cit\t989\foo\target\release\libfoo.rmeta`
Caused by:
  Access is denied. (os error 5)
```

**Pull request merged**